File: client/physics.py
"""A copy of the server's step function, used only to predict my own body.

The server stays authoritative. This exists so my own input shows up on the
next frame instead of one round trip later; anything it gets wrong is pulled
back when the next snapshot arrives.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from playerinput import DOWN, JUMP, LEFT, RIGHT, UP

@dataclass
class Body:
    x: float
    y: float
    vx: float = 0.0
    vy: float = 0.0
    on_ground: bool = False
    on_ladder: bool = False
    frozen: bool = False
    coyote_ms: float = 0.0
    jump_buffer_ms: float = 0.0
    jump_held: bool = False

    def copy(self) -> "Body":
        return Body(**vars(self))


class Physics:
    def __init__(self, game_map, tuning: dict[str, Any]) -> None:
        self.map = game_map
        self.tuning = tuning
        self.solid_by_index = [entry["solid"] for entry in game_map.palette]
        self.ladder_by_index = [entry.get("ladder", False) for entry in game_map.palette]

    def is_solid(self, col: int, row: int) -> bool:
        if col < 0 or row < 0 or col >= self.map.width or row >= self.map.height:
            return True
        return self.solid_by_index[self.map.solid[row][col]]

    def on_ladder(self, x: float, y: float) -> bool:
        """True when the body's middle is over a ladder tile."""
        size = self.map.tile_size
        col = int((x + self.tuning["player_width"] / 2) // size)
        row = int((y + self.tuning["player_height"] / 2) // size)
        if col < 0 or row < 0 or col >= self.map.width or row >= self.map.height:
            return False
        return self.ladder_by_index[self.map.fg[row][col]]

    def overlaps(self, x: float, y: float) -> bool:
        size = self.map.tile_size
        left = int(x // size)
        right = int((x + self.tuning["player_width"] - 1) // size)
        top = int(y // size)
        bottom = int((y + self.tuning["player_height"] - 1) // size)
        for row in range(top, bottom + 1):
            for col in range(left, right + 1):
                if self.is_solid(col, row):
                    return True
        return False

    def step(self, body: Body, mask: int, dt_ms: float) -> None:
        if body.frozen:
            body.vx = body.vy = 0.0
            return

        tuning = self.tuning
        dt = dt_ms / 1000
        wants_jump = bool(mask & JUMP)

        # A ladder suspends gravity while the body is on it; jumping lets go.
        body.on_ladder = self.on_ladder(body.x, body.y) and not wants_jump
        if body.on_ladder:
            body.jump_held = wants_jump
            climb = (1 if mask & DOWN else 0) - (1 if mask & UP else 0)
            body.vy = climb * tuning["climb_speed"]
            direction = (1 if mask & RIGHT else 0) - (1 if mask & LEFT else 0)
            body.vx = direction * tuning["move_speed"]
            self._move_x(body, body.vx * dt)
            landed = self._move_y(body, body.vy * dt)
            body.on_ground = landed
            body.coyote_ms = tuning["coyote_ms"]
            return

        if wants_jump and not body.jump_held:
            body.jump_buffer_ms = tuning["jump_buffer_ms"]
        else:
            body.jump_buffer_ms = max(0.0, body.jump_buffer_ms - dt_ms)

        if not wants_jump and body.jump_held and body.vy < 0:
            body.vy *= tuning["short_hop_factor"]
        body.jump_held = wants_jump

        direction = (1 if mask & RIGHT else 0) - (1 if mask & LEFT else 0)
        body.vx = direction * tuning["move_speed"]

        if body.jump_buffer_ms > 0 and body.coyote_ms > 0:
            body.vy = tuning["jump_speed"]
            body.jump_buffer_ms = 0.0
            body.coyote_ms = 0.0
            body.on_ground = False

        body.vy = min(body.vy + tuning["gravity"] * dt, tuning["max_fall_speed"])

        self._move_x(body, body.vx * dt)
        landed = self._move_y(body, body.vy * dt)

        body.on_ground = landed
        body.coyote_ms = tuning["coyote_ms"] if landed else max(0.0, body.coyote_ms - dt_ms)

    def _move_x(self, body: Body, delta: float) -> None:
        if delta == 0:
            return
        target = body.x + delta
        if not self.overlaps(target, body.y):
            body.x = target
            return
        # A stair is one tile high, so walking into one should climb it rather
        # than stop dead. Only from the ground, so it cannot be used mid-jump.
        if body.on_ground:
            for lift in range(1, int(self.tuning["step_height"]) + 1):
                if not self.overlaps(target, body.y - lift):
                    body.x = target
                    body.y -= lift
                    return

        step_px = 1 if delta > 0 else -1
        while not self.overlaps(body.x + step_px, body.y):
            body.x += step_px
        body.vx = 0.0

    def _move_y(self, body: Body, delta: float) -> bool:
        if delta == 0:
            return body.on_ground and not self.overlaps(body.x, body.y + 1)
        target = body.y + delta
        if not self.overlaps(body.x, target):
            body.y = target
            return False
        step_px = 1 if delta > 0 else -1
        while not self.overlaps(body.x, body.y + step_px):
            body.y += step_px
        landed = delta > 0
        body.vy = 0.0
        return landed

File: server/src/physics.ts
import { isLadderAt, isSolidTile, map, tuning } from "./map";

export const LEFT = 1;
export const RIGHT = 2;
export const JUMP = 4;
export const UP = 8;
export const DOWN = 16;

export type Body = {
  x: number;
  y: number;
  vx: number;
  vy: number;
  onGround: boolean;
  onLadder: boolean;
  frozen: boolean;
  coyoteMs: number;
  jumpBufferMs: number;
  jumpHeld: boolean;
};

export function spawnBody(at: [number, number]): Body {
  return {
    x: at[0],
    y: at[1],
    vx: 0,
    vy: 0,
    onGround: false,
    onLadder: false,
    frozen: false,
    coyoteMs: 0,
    jumpBufferMs: 0,
    jumpHeld: false,
  };
}

/** Advance one body by a single tick. `mask` is this frame's key bitmask. */
export function step(body: Body, mask: number, dtMs: number): void {
  if (body.frozen) {
    body.vx = 0;
    body.vy = 0;
    return;
  }

  const dt = dtMs / 1000;
  const wantsJump = (mask & JUMP) !== 0;

  // A ladder suspends gravity while the body is on it; jumping lets go.
  body.onLadder =
    isLadderAt(body.x + tuning.player_width / 2, body.y + tuning.player_height / 2) &&
    !wantsJump;
  if (body.onLadder) {
    body.jumpHeld = wantsJump;
    const climb = ((mask & DOWN) !== 0 ? 1 : 0) - ((mask & UP) !== 0 ? 1 : 0);
    body.vy = climb * tuning.climb_speed;
    body.vx =
      (((mask & RIGHT) !== 0 ? 1 : 0) - ((mask & LEFT) !== 0 ? 1 : 0)) * tuning.move_speed;
    moveX(body, body.vx * dt);
    body.onGround = moveY(body, body.vy * dt);
    body.coyoteMs = tuning.coyote_ms;
    return;
  }

  body.jumpBufferMs = wantsJump && !body.jumpHeld ? tuning.jump_buffer_ms : Math.max(0, body.jumpBufferMs - dtMs);

  // Releasing jump early cuts the rise short, which is what makes the height
  // controllable rather than fixed.
  if (!wantsJump && body.jumpHeld && body.vy < 0) {
    body.vy *= tuning.short_hop_factor;
  }
  body.jumpHeld = wantsJump;

  const direction = ((mask & RIGHT) !== 0 ? 1 : 0) - ((mask & LEFT) !== 0 ? 1 : 0);
  body.vx = direction * tuning.move_speed;

  if (body.jumpBufferMs > 0 && body.coyoteMs > 0) {
    body.vy = tuning.jump_speed;
    body.jumpBufferMs = 0;
    body.coyoteMs = 0;
    body.onGround = false;
  }

  body.vy = Math.min(body.vy + tuning.gravity * dt, tuning.max_fall_speed);

  moveX(body, body.vx * dt);
  const landed = moveY(body, body.vy * dt);

  body.onGround = landed;
  body.coyoteMs = landed ? tuning.coyote_ms : Math.max(0, body.coyoteMs - dtMs);
}

function overlaps(x: number, y: number): boolean {
  const size = map.tile_size;
  const left = Math.floor(x / size);
  const right = Math.floor((x + tuning.player_width - 1) / size);
  const top = Math.floor(y / size);
  const bottom = Math.floor((y + tuning.player_height - 1) / size);

  for (let row = top; row <= bottom; row++) {
    for (let col = left; col <= right; col++) {
      if (isSolidTile(col, row)) return true;
    }
  }
  return false;
}

function moveX(body: Body, delta: number): void {
  if (delta === 0) return;
  const target = body.x + delta;
  if (!overlaps(target, body.y)) {
    body.x = target;
    return;
  }

  // A stair is one tile high, so walking into one should climb it rather
  // than stop dead. Only from the ground, so it cannot be used mid-jump.
  if (body.onGround) {
    for (let lift = 1; lift <= tuning.step_height; lift++) {
      if (!overlaps(target, body.y - lift)) {
        body.x = target;
        body.y -= lift;
        return;
      }
    }
  }

  // Walk up to the wall one pixel at a time so the body ends up flush with it.
  const stepPx = Math.sign(delta);
  while (!overlaps(body.x + stepPx, body.y)) {
    body.x += stepPx;
  }
  body.vx = 0;
}

/** Returns true when the move ended with ground under the body. */
function moveY(body: Body, delta: number): boolean {
  if (delta === 0) return body.onGround && !overlaps(body.x, body.y + 1);

  const target = body.y + delta;
  if (!overlaps(body.x, target)) {
    body.y = target;
    return false;
  }

  const stepPx = Math.sign(delta);
  while (!overlaps(body.x, body.y + stepPx)) {
    body.y += stepPx;
  }
  const landedOnFloor = delta > 0;
  body.vy = 0;
  return landedOnFloor;
}
